source/common/secret/secret_manager_impl.cc

```diff
@@ -86,15 +86,42 @@ TlsCertificateConfigProviderSharedPtr SecretManagerImpl::findOrCreateTlsCertific
       removeDynamicSecretProvider(map_key);
     };
 
-    secret_provider = std::make_shared<SdsApi>(
+    secret_provider = std::make_shared<TlsCertificateSdsApi>(
         secret_provider_context.localInfo(), secret_provider_context.dispatcher(),
         secret_provider_context.random(), secret_provider_context.stats(),
         secret_provider_context.clusterManager(), *secret_provider_context.initManager(),
         sds_config_source, config_name, unregister_secret_provider);
     dynamic_secret_providers_[map_key] = secret_provider;
   }
 
-  return secret_provider;
+  return std::dynamic_pointer_cast<TlsCertificateConfigProvider>(secret_provider);
+}
+
+CertificateValidationContextConfigProviderSharedPtr
+SecretManagerImpl::findOrCreateCertificateValidationContextProvider(
```

Comment:
The code is similar for this function and the other.  try to share them by create a 

SdsApiSharedPtr innerFindOrCreate(...,   create_fn);

Then,  each function will provide its creation_function to create proper object,

Reply:
Done.